extensions/amp-autocomplete/validator-amp-autocomplete.protoascii

```diff
@@ -0,0 +1,102 @@
+#
+# Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the license.
+#
+
+tags: {  # amp-autocomplete
+  html_format: AMP
+  tag_name: "SCRIPT"
+  extension_spec: {
+    name: "amp-autocomplete"
+    version: "0.1"
+    version: "latest"
+  }
+  attr_lists: "common-extension-attrs"
+}
+tags: {  # <amp-autocomplete>
+  html_format: AMP
+  tag_name: "AMP-AUTOCOMPLETE"
+  requires_extension: "amp-autocomplete"
+  requires_extension: "amp-form"
```

Comment:
I might recommend against this line, though it's reasonable either way.

My thought is that this tag already mandates that it's ancestor is `form` which requires the `amp-form` extension itself. So, transitively, `amp-form` is required anyway without this `requires_extension` field.

Let's say a user creates the page correctly, but forgets the `amp-form` extension code:

```
<form ...>
  <amp-autocomplete ...>
  </amp-autocomplete>
</form>
```

If `amp-autocomplete` requires `amp-form`, then the validator will create two errors about missing `amp-form`, one when it processes the `<form>` element, and another when processing `<amp-autocomplete>`. If `amp-autocomplete` does not require `amp-form`, then only one error will be emitted. Fewer errors is probably better as long as the set is comprehensive.

I don't feel strongly however, so this is just a suggestion.

Reply:
This is a useful point. To confirm, the validation rules should be a succinct set of rules that minimally cover the requirements. In this case the `requires_extension` is a redundant attribute because its inclusion in the document is made mandatory by the same check when `form` becomes included.

What about the case where a user creates `amp-autocomplete` without a `form` ancestor? Then there is one error to include the tag, and the second error about the missing `amp-form` extension does not appear until after an additional pass with the new change. Is that a standard flow or is it better to be straight forward about displaying known requirements early on?

Reply:
Agreed with your assessment. I don't think there is a single standard here, and I'm happy either way.